### ISSUE TITLE
how to transfer to another account without using api.tx.balances.transfer?

### ISSUE THREAD

**Issue by Wangmmx** [polkadot-js/api#1421]
I want to know the extrinsic body or tx body to sign, so that I can send and broadcast in my local test, plz help me ,thx a lot

**Comment by jacogr** [comment]
This could be useful.

https://polkadot.js.org/api/start/FAQ.html#i-would-like-to-sign-transactions-offline

**Comment by Wangmmx** [comment]
> This could be useful.
> 
> https://polkadot.js.org/api/start/FAQ.html#i-would-like-to-sign-transactions-offline

thanks a lot!  I found the way to make signature offline, but how to determine the params of tx body? I could not work it out >-<

**Comment by jacogr** [comment]
The body is a payload, which is generated based on the on-chain extrinsic version.e.g. for v3 extrinsics, it looks like this 

https://github.com/polkadot-js/api/blob/master/packages/types/src/primitive/Extrinsic/v3/ExtrinsicPayload.ts

That is exactly what the `RawSigner` in the tools repo returns by hooking into signAndSend, filling in all the blanks with regards to versions, the extrinsic era and applicable hashes.

**Comment by Wangmmx** [comment]
> RawSigner

I got an example of payload body at https://github.com/polkadot-js/api/blob/master/packages/types/src/primitive/Extrinsic/ExtrinsicPayload.spec.ts 
like this
![image](https://user-images.githubusercontent.com/22836587/65671792-78460f00-e07a-11e9-8635-9a8ab198168f.png)

and I created a payload and signed , I got the data and signature after `createType('ExtrinsicPayload', payload, { version: payload.version }).sign(pair)`

 but it was invalid in the tools, did I do this right?

**Comment by jacogr** [comment]
I'm assuming . by tools, you mean those in the apps UI.

So what the signing does is to sign the bytes payload, so something like this should work -

```js
const signable = createType('ExtrinsicPayload', payload, { version: payload.version });
const { signature } = signable.sign(pair);
const verifyData = signable.toU8a(true);

// this is weird though, need to log an issue in common to take in both hex & Uint8Array
// inputs for both message and signature...
pair.verify(verifyData, hexToU8a(signature));
```

**Comment by Wangmmx** [comment]
> I'm assuming by tools, you mean those in the apps UI.
> 
> So what the signing does is to sign the bytes payload, so something like this should work -
> 
> ```js
> // create the signable payload
> const signable = createType('ExtrinsicPayload', payload, { version: payload.version });
> 
> // get the signature
> const { signature } = signable.sign(pair);
> 
> // this is exactly what is signed
> const verifyData = signable.toU8a(true);
> 
> // this is weird though, need to log an issue in common to take in
> // both hex & Uint8Array inputs for both message and signature...
> pair.verify(verifyData, hexToU8a(signature));
> ```
> 
> If you want something valid to pass to the UI, it should be `signature` and
> 
> ```js
> const message = u8aToHex(verifyData);
> ```

It helped a lot!  
I built a payload and signed it successfully, the result of verifying in tools was correct
now I am going to generate extrinsic and try to submit it, thank u

**Comment by jacogr** [comment]
Please indicate here if you need any additional pointers.

**Comment by Wangmmx** [comment]
hello,I got a problem when I created submittable:
I used createSubmittable function in 
https://github.com/polkadot-js/api/blob/ad4a6e63ce890d875a29eebf93c6df373962849d/packages/api/src/submittable/createSubmittable.ts
 but it didn't work
did I need to create a submittable and sign?
here are my steps to make a transaction offline:
1.create a payload
`let ExtrinsicPayload = createType('ExtrinsicPayload', payload, { version: payload.version });`
2.sign the payload created just now
` let signed = ExtrinsicPayload.sign(Pair)`
3.create a submittable and sign
4.create an Extrinsic
` let extrinsicCreation = createType('Extrinsic', signedSubmittable)`
5.submit the result
` let submit = await api.rpc.author.submitExtrinsic(extrinsicCreation);`

did I made the correct order? how can I create a submittable at step3?
sorry for my poor English, it is difficult to understand >-< 
thank u very much!

**Comment by jacogr** [comment]
To sign and send a transaction, there are a couple of things that needs to be done -

1. Create the transaction
2. Create signing payload
3. Sign the payload
4. Inject payload data & signature into the transaction
5. Send the transaction

Let's walk through -

```js
// PROCESS 1

// create the transaction
const tx = api.tx.balances.transfer(AN_ADDR, 12345);

// create the payload
const payload = api.createType('ExtrinsicPayload', {
  method: tx.toHex(),
  era: ... // mortal era or empty for immortal
  nonce: ...
  tip: ...
  specVersion: api.runtimeVersion.specVersion,
  genesisHash: api.genesisHash,
  blockHash: ... // blockhash for the era (or genesisHash if immortal)
}, { version: 3 });

// serialize, assuming we are taking it somewhere else
const hexPayload = payload.toHex();
```

Ok, so the first 2 steps done. Now we get to the signing.

```js
// PROCESS 2

// re-create the payload from the hex (we are assuming this is "somewhere else now")
const payload = api.createType('ExtrinsicPayload', hexPayload, { version: 3 });

// sign the actual payload
const signature = pair.sign(payload.toU8a(true));

// convert sign back to hex to transfer back to the original
const sigHex = u8aToHex(signature);
```

At this point we have a valid payload and a valid signature - inject all the data and get the stuff sent.

```js
// PROCESS 1

// inject the signature and original payload we have
tx.addSignature(address, sigHex, payload);

// send the transaction now that is is signed
const hash = await tx.send();
```

**Comment by Wangmmx** [comment]
> To sign and send a transaction, there are a couple of things that needs to be done -
> 
> 1. Create the transaction
> 2. Create signing payload
> 3. Sign the payload
> 4. Inject payload data & signature into the transaction
> 5. Send the transaction
> 
> Let's walk through -
> 
> ```js
> // PROCESS 1
> 
> // create the transaction
> const tx = api.tx.balances.transfer(AN_ADDR, 12345);
> 
> // create the payload
> const payload = api.createType('ExtrinsicPayload', {
>   method: tx.toHex(),
>   era: ... // mortal era or empty for immortal
>   nonce: ... // nonce for the acocunt
>   tip: ... // any tip (leave empty for 0)
>   specVersion: api.runtimeVersion.specVersion,
>   genesisHash: api.genesisHash,
>   blockHash: ... // blockhash for the era (or genesisHash if immortal)
> }, { version: 3 });
> 
> // serialize, assuming we are taking it somewhere else
> const hexPayload = payload.toHex();
> ```
> 
> Ok, so the first 2 steps done. Now we get to the signing.
> 
> ```js
> // PROCESS 2
> 
> // re-create the payload from the hex (we are assuming this is "somewhere else now")
> const payload = api.createType('ExtrinsicPayload', hexPayload, { version: 3 });
> 
> // sign the actual payload
> const signature = pair.sign(payload.toU8a(true));
> 
> // convert signature back to hex to transfer back to the original
> const sigHex = u8aToHex(signature);
> 
> // we also want the address of the signer to put inside process 1
> const address = pair.address;
> ```
> 
> At this point we have a valid payload and a valid signature - inject all the data and get the stuff sent.
> 
> ```js
> // PROCESS 1
> 
> // inject the address, signature and original payload we have
> tx.addSignature(address, sigHex, payload);
> 
> // send the transaction now that is is signed
> const hash = await tx.send();
> ```

You are so patient and kind！

is there any other way to create the transaction? 
I don't want to connect with the websocket or others, so I try to create transaction without using api.tx.balances.transfer, I read the origin code times and selected the useful parts all the time , hah

**Comment by amaurymartiny** [comment]
Hey @Wangmmx, here's a way to do it without ever touching the `api` object or any providers:

https://github.com/amaurymartiny/offline/blob/master/src/extrinsic.ts#L15-L21

**Comment by Wangmmx** [comment]
> Hey @Wangmmx, here's a way to do it without ever touching the `api` object or any providers:
> 
> https://github.com/amaurymartiny/offline/blob/master/src/extrinsic.ts#L15-L21

thanks a lot! I will go see it now!

**Comment by jacogr** [comment]
@Wangmmx I would strongly suggest that in `PROCESS 1` you use the API - you need to ensure that you have the metadata for the chain and it is properly injected. 

(The static metadata is randomly updated and specific to a chain, it should never be used, i.e. it is based on a Substrate master dev chain, which is vastly different from e.g. Kusama)

In your offline signing, e.g. `PROCESS 2`, you don't need any API.

**Comment by Wangmmx** [comment]
> @Wangmmx I would strongly suggest that in `PROCESS 1` you use the API - you need to ensure that you have the metadata for the chain and it is properly injected.
> 
> (The static metadata is randomly updated and specific to a chain, it should never be used, i.e. it is based on a Substrate master dev chain, which is vastly different from e.g. Kusama)
> 
> In your offline signing, e.g. `PROCESS 2`, you don't need any API.

Oh, I understand what you mean, thank you very much, I will try again and use API in PROCESS

**Comment by jacogr** [comment]
Sorry if I wasn't 100% clear in my `PROCESS {1,2}` descriptions :) If was meant to be "connected process" and "offline non-api process".

The metadata is tricky, since the whole API is decorated based on the metadata. This means it has all the indexes of the specific modules and calls decorated from the metadata. So unlike some other blockchains, these things aren't static and can only be 100% known when actually connected. (Luckily in the signing, we only need the bytes, so it all can be offline)

**Comment by Wangmmx** [comment]
> Sorry if I wasn't 100% clear in my `PROCESS {1,2}` descriptions :) If was meant to be "connected process" and "offline non-api process".
> 
> The metadata is tricky, since the whole API is decorated based on the metadata. This means it has all the indexes of the specific modules and calls decorated from the metadata. So unlike some other blockchains, these things aren't static and can only be 100% known when actually connected. (Luckily in the signing, we only need the bytes, so it all can be offline)

I see, I connected provider API and had another try.

 I used api.tx.balances.transfer() to create a transaction, and then I created a payload, signed it, but  I got an error when sending the tx. Comparing with the signature after api.tx.balances.transfer().sign, I found the two signatures were different, so I guess signature which I signed payload was not correct.

 The problem is not solved yet. Could you please help me? Here are my codes

`  const from = '5HV55pUNEmaU4Trgb7MYLVTGCd5BArkins9Z79ncC658ojzk';
    const to = '5GXvoYbSd2VZeabSY63mGdGJc6bDrmZg5mTz78ueEhKMAivm';

    const signedBlock = await api.rpc.chain.getBlock();
    const currentHeight = signedBlock.block.header.number;
    const era = createType('ExtrinsicEra', { current: currentHeight, period: 10 });
    const blockHash = signedBlock.block.header.hash;
    const nonce = await api.query.system.accountNonce(from);
    const genesisHash = await api.genesisHash;
    const specVersion = await api.runtimeVersion.specVersion;

    const tx = api.tx.balances.transfer(to, 1);

    let payload = createType('ExtrinsicPayload', {
        blockHash: blockHash.toHex(),
        era: era.toHex(),
        genesisHash: genesisHash.toHex(),
        method: tx.toHex(),
        nonce: nonce.toHex(),
        specVersion: specVersion.toHex(),
        tip: ''
    }, { version : 3} );

    const signature = pair.sign(payload.toU8a(true));
    const sigHex = u8a.u8aToHex(signature);
    console.log(sigHex)

    tx.addSignature(from, sigHex, payload);
    const hash = await tx.send();
    console.log(hash)`


and the error is 1010: Invaslid Transaction(0)
![image](https://user-images.githubusercontent.com/22836587/66549543-11dee780-eb76-11e9-9f18-63b87f7bb782.png)

**Comment by jacogr** [comment]
I would suggest using ed25519 accounts as a start - sr25519 has non-deterministic signatures. (So if you sign in two different places, you won't be able to compare). However, you are right `(0)` is invalid signature.

`tip: ''` I would drop. (when not supplied it will end up as `0`)

The only one you need to call `toHex()` on explicitly is the `method`.

Some additional comments, neither of these are promises -

```
const genesisHash = await api.genesisHash;
const specVersion = await api.runtimeVersion.specVersion;
```

Then... which node are you passing this through? For instance, Alexander has extrinsic version 1, Kusama has version 3. (And substrate master is on 3 as well)

**Comment by Wangmmx** [comment]
> I would suggest using ed25519 accounts as a start - sr25519 has non-deterministic signatures. (So if you sign in two different places, you won't be able to compare). However, you are right `(0)` is invalid signature.
> 
> `tip: ''` I would drop. (when not supplied it will end up as `0`)
> 
> The only one you need to call `toHex()` on explicitly is the `method`.
> 
> Some additional comments, neither of these are promises -
> 
> ```
> const genesisHash = await api.genesisHash;
> const specVersion = await api.runtimeVersion.specVersion;
> ```
> 
> Then... which node are you passing this through? For instance, Alexander has extrinsic version 1, Kusama has version 3. (And substrate master is on 3 as well)
> 
> EDIT: `api.extrinsicVersion` (non-promise, just a getter like genesisHash & version) has the extrinsic version as detected on-chain.

thanks for suggestions!

I used ed25519 to generate the pair
I drop the tip: ''

Yes you are right, I should select Alexander version, which is 1, and the result of api.extrinsicVersion is 1, I fix my code but the problem is still exist, the two signatures was not the same, it makes me really upset

**Comment by Wangmmx** [comment]
> Did you remove the await for `genesisHash` & `specVersion` as well?
> 
> Additionally what happens when you do -
> 
> * also try the `payload.sign(pair)` shortcut (gets rid of the `toU8a()`..., less boilerplate)
> * does `console.log(JSON.stringify(payload))` match up with your inputs?

yes I removed the await for `genesisHash` & `specVersion` 

I tried payload.sign(pair) just now,  it was the same as pair.sign(payload.toU8a(true))

I found the problem,  after creating payload, some datas (payload.genesisHash and specVersion )were changed to 0,  but in version 3 , datas are right

**Comment by jacogr** [comment]
So `genesisHash` & `specVersion` does not apply to v1 (the first was a v2 change, the second a v3 change). So they actually are not serialized when signing, so all ok there.

**Comment by Wangmmx** [comment]
> So `genesisHash` & `specVersion` does not apply to v1 (the first was a v2 change, the second a v3 change). So they actually are not serialized when signing, so all ok there.

but I am using Alexander, version should be 1, version 1 got the (0) error, using api.tx.balances.transfer(to,1).sign(pair, nonce) is working

**Comment by Wangmmx** [comment]
> ```js
> const { signature } = payload.sign(pair);
> ```

thank you for telling me this, I checked and it was alright

**Comment by jacogr** [comment]
So I'm assuming you passed in -

```js
let payload = createType('ExtrinsicPayload', {
    ...
}, { version: 1 } );
```

Then yes, it will be 1.

**Comment by Wangmmx** [comment]
> So I'm assuming you passed in -
> 
> ```js
> let payload = createType('ExtrinsicPayload', {
>     ...
> }, { version: 1 } );
> ```
> 
> Then yes, it will be 1.

I can list the data after signing
1.data after api.tx.balances.transfer(to,1).sign(pair, nonce)
` nonce 0x000000000000000e
era 0x00
genesisHash 0xdcd1346701ca8396496e52aa2785b1748deb6db09551b72159dcb3e08991025b
method 0x0300ffc5b423eea8b3438b5d16f8d34c2a54f2b7a8b65194d5d88a5fda9966de8b286a04
`

2.data after signing payload
`nonce 0x000000000000000e
genesisHash 0x0000000000000000000000000000000000000000000000000000000000000000
specVersion 0x00000000
blockHash 0x699053f7e975eb05892cd122cebf9e2474c9b9220f3c147cd17efa5296d9f960
era 0x8300
method 0x94010300ffc5b423eea8b3438b5d16f8d34c2a54f2b7a8b65194d5d88a5fda9966de8b286a04`

it seems different from each other, so can I use version 1 to create a payload and sign correctly?
sorry for bothering you all the time, I am confused and anxious

**Comment by Wangmmx** [comment]
> So I'm assuming you passed in -
> 
> ```js
> let payload = createType('ExtrinsicPayload', {
>     ...
> }, { version: 1 } );
> ```
> 
> Then yes, it will be 1.

I have test dots in Alexander, so I have to use version 1,  ws url is wss://poc3-rpc.polkadot.io

**Comment by jacogr** [comment]
The signing is not a simple thing, which is why we have the API methods setup the way they are - and generally would recommend using a signer should you wish to inject yourself into the signing process.  (As per the offline sample in tools) 

You never should have to use `createType` anywhere, the fact that you does means you are -

- understanding the data structures your are working with
- trying to do something out of the realm of API support

TL;DR It looks complex, because it is - the API wraps and abstracts all this away from the users for a reason.

In the above, the payload data actually looks wrong. The method should not be prefixed for signing. You are also not comparing the two 100% since in the one case you pass an immortal and in the other case a immortal. (So would swap the payload to immortal until you have it aligned)

So do the following to debug -

For the api tx:

- `tx = api.tx.balance.transfer(...).sign(pair, { nonce })`
- (straight after) `console.log(tx.toHex())`

For the payload:

- `tx = api.tx.balance.transfer(...)`
- `payload = createType(...)`
- `payload.sign(...)`
- `tx.addSignature(...)`
- (after all that) `console.log(tx.toHex())`

Manually decoding both of those into their various parts should show where the issue lies - wo please post them here

**Comment by Wangmmx** [comment]
> The signing is not a simple thing, which is why we have the API methods setup the way they are - and generally would recommend using a signer should you wish to inject yourself into the signing process. (As per the offline sample in tools)
> 
> You never should have to use `createType` anywhere, the fact that you does means you are -
> 
> * understanding the data structures your are working with
> * trying to do something out of the realm of API support
> 
> TL;DR It looks complex, because it is - the API wraps and abstracts all this away from the users for a reason.
> 
> In the above, the payload data actually looks wrong. The method should not be prefixed for signing. You are also not comparing the two 100% since in the one case you pass an immortal and in the other case a immortal. (So would swap the payload to immortal until you have it aligned)
> 
> So do the following to debug -
> 
> For the api tx:
> 
> * `tx = api.tx.balance.transfer(...).sign(pair, { nonce })`
> * (straight after) `console.log(tx.toHex())`
> 
> For the payload:
> 
> * `tx = api.tx.balance.transfer(...)`
> * `payload = createType(...)`
> * `payload.sign(...)`
> * `tx.addSignature(...)`
> * (after all that) `console.log(tx.toHex())`
> 
> Manually decoding both of those into their various parts should show where the issue lies - wo please post them here

For the api tx:
0x210281ffefc28334ca379220760f593f48cfc918172e0b707444bcbea5340d8445609ceff1e68d989a67487a2d6ebe2ae2ecac19e788998a864133efdf4e2be4c2f8a1da2ba80dfc2e97cf7cb364461e347069cffcc7a4b667632ad1b9b42c64d30adb0e38000300ffc5b423eea8b3438b5d16f8d34c2a54f2b7a8b65194d5d88a5fda9966de8b286a04


For the payload:
0x250281ffefc28334ca379220760f593f48cfc918172e0b707444bcbea5340d8445609cef00369b74eb6060a0e413293f3b5150aae69daaf61f2cc3532e310a9912387bdc4073582e7862f7fb940ab4adea6bafccd7f73fa8b8c6c813ec4844a1e24ecd0d3883000300ffc5b423eea8b3438b5d16f8d34c2a54f2b7a8b65194d5d88a5fda9966de8b286a04


I tried decode the tx hex, the two signatures have different signatures and era

**Comment by jacogr** [comment]
Yes, they will have different era, since you sign without one in the first `tx.sign()`. (So I would suggest removing the era in the payload part, just until this is sorted)

**Comment by Wangmmx** [comment]
> Yes, they will have different era, since you sign without one in the first `tx.sign()`. (So I would suggest removing the era in the payload part, just until this is sorted)

remove era:
0x210281ffefc28334ca379220760f593f48cfc918172e0b707444bcbea5340d8445609cefa345bfaf6c2b0e926a759f7529ddecdb76911fedc45888f3dee53519226fd4e190976f775c033d420cf47d221c5d223ce82440d65bd2df4f2776bec087e8d80138000300ffc5b423eea8b3438b5d16f8d34c2a54f2b7a8b65194d5d88a5fda9966de8b286a04

only different in signature, what are tx.sign() includes? should I remove other?

**Comment by jacogr** [comment]
Cool. 

Can you output the following for both as well?

```js
console.log('tx api', JSON.stringify({ data: tx1.data, era: tx1.era, method: tx1.method, nonce: tx1.nonce, signer: tx1.signer }));
console.log('tx pay', JSON.stringify({ data: tx2.data, era: tx2.era, method: tx2.method, nonce: tx2.nonce, signer: tx2.signer }));
```

(When removing the era, just make 100% sure you pass `blockHash: api.genesisHash` into the payload)

**Comment by Wangmmx** [comment]
> Cool.
> 
> Can you output the following for both as well?
> 
> ```js
> console.log('tx api', JSON.stringify({ data: tx1.data, era: tx1.era, method: tx1.method, nonce: tx1.nonce, signer: tx1.signer }));
> console.log('tx pay', JSON.stringify({ data: tx2.data, era: tx2.era, method: tx2.method, nonce: tx2.nonce, signer: tx2.signer }));
> ```
> 
> (When removing the era, just make 100% sure you pass `blockHash: api.genesisHash` into the payload)

tx api {"data":{"0":255,"1":197,"2":180,"3":35,"4":238,"5":168,"6":179,"7":67,"8":139,"9":93,"10":22,"11":248,"12":211,"13":76,"14":42,"15":84,"16":242,"17":183,"18":168,"19":182,"20":81,"21":148,"22":213,"23":216,"24":138,"25":95,"26":218,"27":153,"28":102,"29":222,"30":139,"31":40,"32":106,"33":4},"era":{"ImmortalEra":"0x00"},"method":{"callIndex":"0x0300","args":{"dest":"5GXvoYbSd2VZeabSY63mGdGJc6bDrmZg5mTz78ueEhKMAivm","value":1}},"nonce":14,"signer":"5HV55pUNEmaU4Trgb7MYLVTGCd5BArkins9Z79ncC658ojzk"}



tx pay {"data":{"0":255,"1":197,"2":180,"3":35,"4":238,"5":168,"6":179,"7":67,"8":139,"9":93,"10":22,"11":248,"12":211,"13":76,"14":42,"15":84,"16":242,"17":183,"18":168,"19":182,"20":81,"21":148,"22":213,"23":216,"24":138,"25":95,"26":218,"27":153,"28":102,"29":222,"30":139,"31":40,"32":106,"33":4},"era":{"ImmortalEra":"0x00"},"method":{"callIndex":"0x0300","args":{"dest":"5GXvoYbSd2VZeabSY63mGdGJc6bDrmZg5mTz78ueEhKMAivm","value":1}},"nonce":14,"signer":"5HV55pUNEmaU4Trgb7MYLVTGCd5BArkins9Z79ncC658ojzk"}

These two are the same

**Comment by jacogr** [comment]
Can you just post how you are creating the payload atm?

**Comment by Wangmmx** [comment]
> Can you just post how you are creating the payload atm?

    let payload = createType('ExtrinsicPayload', {
        blockHash: blockHash.toHex(),
       // era: era,
        genesisHash: genesisHash,
        method: tx.toHex(),
        nonce: nonce.toHex(),
        specVersion: specVersion,
      //  tip: ''
    }, { version : 1} );

version 1 or 3 will not work

**Comment by jacogr** [comment]
`blockHash: blockHash.toHex(),` - you need to pass the `genesisHash` there when you don't specify and era, as mentioned in https://github.com/polkadot-js/api/issues/1421#issuecomment-540516753 as well as https://github.com/polkadot-js/api/issues/1421#issuecomment-540512563

**Comment by Wangmmx** [comment]
> `blockHash: blockHash.toHex(),` - you need to pass the `genesisHash` there when you don't specify an era, as mentioned in [#1421 (comment)](https://github.com/polkadot-js/api/issues/1421#issuecomment-540516753) as well as [#1421 (comment)](https://github.com/polkadot-js/api/issues/1421#issuecomment-540512563)

sorry I missed this one, is it right now?
let payload = createType('ExtrinsicPayload', {
        blockHash: genesisHash,
       // era: era,
        genesisHash: genesisHash,
        method: tx.toHex(),
        nonce: nonce.toHex(),
        specVersion: specVersion,
      //  tip: ''
    }, { version : 1} );

still (0) error

**Comment by jacogr** [comment]
The above is more-or-less I believe what I would have done, apart from `nonce` which incurs additional costs since we convert to-and-from the compact representation. (These are always problematic since for JSON it is BE and for SCALE they are LE)

**Comment by Wangmmx** [comment]
> The above is more-or-less I believe what I would have done, apart from `nonce` which incurs additional costs since we convert to-and-from the compact representation. (These are always problematic since for JSON it is BE and for SCALE they are LE)

these are my test data:
    const keyring = new Keyring({type: 'ed25519'});
    const pair = keyring.addFromUri('obscure pen impulse report glow soon bacon turn find tomorrow journey mimic')
    const from = '5HV55pUNEmaU4Trgb7MYLVTGCd5BArkins9Z79ncC658ojzk';
    const to = '5GXvoYbSd2VZeabSY63mGdGJc6bDrmZg5mTz78ueEhKMAivm';

and transfer amount is 1

when the tx is 0x210281ffefc28334ca379220760f593f48cfc918172e0b707444bcbea5340d8445609ceff1e68d989a67487a2d6ebe2ae2ecac19e788998a864133efdf4e2be4c2f8a1da2ba80dfc2e97cf7cb364461e347069cffcc7a4b667632ad1b9b42c64d30adb0e38000300ffc5b423eea8b3438b5d16f8d34c2a54f2b7a8b65194d5d88a5fda9966de8b286a04 

it is correct , if you can submit tx successfully, please tell me the details, thanks again!

**Comment by jacogr** [comment]
Ok, this stuff really was not meant to be used this way. I would really appreciate it is people with explicit offline signing requirements can actually  log these requirements. So we can actually address them properly.

So this works -

```js
const Alice = keyring.getPair('5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY');
const tx = api.tx.balances.transfer('5GNJqTPyNqANBkUVMN1LPPrxXnFouWXoe2wNSmmEoLctxiZY', 12345678);
const nonce = await api.query.system.accountNonce('5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY');

// create the payload
const signer = api.createType('SignerPayload', {
  method: tx,
  nonce,
  genesisHash: api.genesisHash,
  blockHash: api.genesisHash,
  runtimeVersion: api.runtimeVersion,
  version: api.extrinsicVersion
});

const { signature } = api.createType('ExtrinsicPayload', signer.toPayload(), { version: api.extrinsicVersion }).sign(Alice);

tx.addSignature(Alice.address, signature, signer.toPayload());

try {
  console.log(await tx.send());
} catch (error) {
  console.error(error);
}
```

(Which just mimicks what the API does internally. The API does not seem to handle `ExtrinsicPayload` properly, only the extended `SignerPayload` variant - which works nicely with the existing signing capabilities)

I would strongly urge you to explicitly log the exact requirements in a seperate issue so you don't have to do this. It is really tough tyring to figure out what people are trying to do and solve issues on the API where it is used in a way where we have not had any requirements for in the past.

**Comment by Wangmmx** [comment]
> Ok, this stuff really was not meant to be used this way. I would really appreciate it is people with explicit offline signing requirements can actually log these requirements. So we can actually address them properly.
> 
> So this works -
> 
> ```js
> const Alice = keyring.getPair('5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY');
> const tx = api.tx.balances.transfer('5GNJqTPyNqANBkUVMN1LPPrxXnFouWXoe2wNSmmEoLctxiZY', 12345678);
> const nonce = await api.query.system.accountNonce('5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY');
> 
> // create the payload
> const signer = api.createType('SignerPayload', {
>   method: tx,
>   nonce,
>   genesisHash: api.genesisHash,
>   blockHash: api.genesisHash,
>   runtimeVersion: api.runtimeVersion,
>   version: api.extrinsicVersion
> });
> 
> const { signature } = api.createType('ExtrinsicPayload', signer.toPayload(), { version: api.extrinsicVersion }).sign(Alice);
> 
> tx.addSignature(Alice.address, signature, signer.toPayload());
> 
> try {
>   console.log(await tx.send());
> } catch (error) {
>   console.error(error);
> }
> ```
> 
> (Which just mimicks what the API does internally. The API does not seem to handle `ExtrinsicPayload` properly, only the extended `SignerPayload` variant - which works nicely with the existing signing capabilities)
> 
> I would strongly urge you to explicitly log the exact requirements in a seperate issue so you don't have to do this. It is really tough tyring to figure out what people are trying to do and solve issues on the API where it is used in a way where we have not had any requirements for in the past.

it works! thank you very much! I am so happy now hah!
and I found account '5GrwvaEF5zXb26Fz9rcQpDWS57CtERHpNehXCPcNoHGKutQY'  is not active, It's ok I will create another address 
thanks again!

**Comment by jacogr** [comment]
That address is the standard `//Alice` dev address - the above was typed up against a dev chain on the JS console in the UI.

**Comment by jacogr** [comment]
I am closing this - please, please, please add your explict signing requirements as to the full flow that works in your case to #1461

**Comment by jacogr** [comment]
@Wangmmx I saw why it didn't work initially - the `method` in the `ExtrinsicPayload` is quite different from how an actual Call or Extrinsic is defined. The `SignerPayload` handles this conversion transparently to convert between the types. 

A simple `extrinsic.toHex()` as an input does not work for the `ExtrinsicPayload` - however a `{ method: tx.method.toHex() }` _should_ work on the `ExtrinsicPayload` types with no jumping to `SignerPayload` in-between.

The types here with the signing is _really_ tricky...

**Comment by Rakhymzhan** [comment]
> This could be useful.
> 
> https://polkadot.js.org/api/start/FAQ.html#i-would-like-to-sign-transactions-offline

Hey, can you help with steps to sign and send online,

**Comment by MrRakhim** [comment]
> Hey @Wangmmx, here's a way to do it without ever touching the `api` object or any providers:
> 
> https://github.com/amaurymartiny/offline/blob/master/src/extrinsic.ts#L15-L21

can you update the link above please? because it is not available now
